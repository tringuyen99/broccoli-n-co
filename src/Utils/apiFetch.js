import axios from 'axios';
import isEmpty from 'lodash/isEmpty';

/**
 * Fetch wrapper for our backend api
 * @param url
 * @param method
 * @param formData
 * @returns {Promise<Response | never>}
 */
export function apiFetch(url, method, formData) {
    let apiUrl = url;
    const data = {
        method,
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
    };

    if (!isEmpty(formData)) {
        if (method === 'GET') {
            apiUrl +=
                '?' +
                Object.keys(formData)
                    .map(key => key + '=' + formData[key])
                    .join('&');
        } else {
            data.data = formData;
        }
    }

    return axios(apiUrl, data)
        .catch(error => {
            console.log(error.response)
            throw error.response;
        });
}