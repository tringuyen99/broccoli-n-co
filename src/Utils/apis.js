export const invitationEmailAPI = process.env.REACT_APP_BACKEND_API + "/fakeAuth";