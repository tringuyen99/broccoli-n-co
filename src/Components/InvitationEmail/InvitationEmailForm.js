import { useRef, useEffect, useState } from 'react';
import Proptypes from 'prop-types';
import * as Yup from 'yup';

// Components
import { Formik } from 'formik';
import { Container, Form, Button } from 'react-bootstrap';

// Utils and API
import { apiFetch } from '../../Utils/apiFetch';
import { invitationEmailAPI } from '../../Utils/apis';

import './InvitationEmailForm.scss';

// Schema for yup
const validationSchema = Yup.object().shape({
    name: Yup.string()
        .min(3, '*Names must have at least 3 characters')
        .required('*Name is required'),
    email: Yup.string()
        .email('*Must be a valid email address')
        .required('*Email is required'),
    confirmEmail: Yup.string()
        .email('*Must be a valid email address')
        .required('*Please confirm your email')
        .oneOf([Yup.ref('email'), null], '*Emails must match'),
});

function InvitationEmailForm(props) {
    const isUnmounted = useRef(false);
    const [submitError, setSubmitError] = useState("");

    useEffect(() => {
        return () => {
            isUnmounted.current = true;
        }
    }, []);

    return (
    <Container fluid className="justify-content-center text-center">
        <h1 className="display-6 fw-bold">Please fill in this form</h1>
        <p className="fw-light">Broccoli&Co</p>
        <Formik 
            initialValues={{ name:"", email:"", confirmEmail:""}}
            validationSchema={validationSchema}
            onSubmit={(values, {setSubmitting, resetForm}) => {
                
                const submittedData = {
                    "name": values.name,
                    "email": values.email,
                }
                apiFetch(invitationEmailAPI, 'POST', submittedData)
                    .then(() => {
                        props.onSuccess();
                    })
                    .catch(error => {
                        setSubmitError(error.data.errorMessage);
                    })
                    .finally(() => {
                        if (!isUnmounted.current) {
                            setSubmitting(false);
                            resetForm();
                        }
                    })
            }}
        >
            {( {values,
                errors,
                touched,
                handleChange,
                handleBlur,
                handleSubmit,
                isValid,
                isSubmitting }) => (
                <Form onSubmit={handleSubmit} className="mx-auto">
                    <Form.Group controlId="formName"className="mb-4">
                        <Form.Label>Name :</Form.Label>
                        <Form.Control
                            className={touched.name && errors.name ? "border border-warning" : "border border-dark"}
                            type="text"
                            name="name"
                            placeholder="Full Name"
                            onChange={handleChange}
                            onBlur={handleBlur}
                            value={values.name}
                        />
                        {touched.name && errors.name ? (
                            <div className="error-message">{errors.name}</div>): null
                        }
                    </Form.Group>
                    <Form.Group controlId="formEmail" className="mb-4">
                        <Form.Label>Email :</Form.Label>
                        <Form.Control
                            className={touched.email && errors.email ? "border border-3 border-warning" : "border border-dark"}
                            type="text"
                            name="email"
                            placeholder="Email"
                            onChange={handleChange}
                            onBlur={handleBlur}
                            value={values.email}
                        />
                        {touched.email && errors.email ? (
                            <div className="error-message">{errors.email}</div>): null
                        }
                    </Form.Group>
                    <Form.Group controlId="formConfirmEmail" className="mb-5">
                        <Form.Label>Confirm Email :</Form.Label>
                        <Form.Control
                            className={touched.confirmEmail && errors.confirmEmail ? "border border-warning" : "border border-dark"}
                            type="text"
                            name="confirmEmail"
                            placeholder="Confirm Email"
                            onChange={handleChange}
                            onBlur={handleBlur}
                            value={values.confirmEmail}
                        />
                        {touched.confirmEmail && errors.confirmEmail ? (
                            <div className="error-message">{errors.confirmEmail}</div>): null
                        }
                    </Form.Group>
                    <div className="d-grid gap-2">
                        {submitError ? (
                            <div className="mb-2 text-danger">{submitError}</div>): null
                        }
                        <Button variant="primary" size="lg" type="submit" disabled={!isValid || isSubmitting}>
                            {isSubmitting? "Submitting" : "Submit"}
                        </Button>
                    </div>
                </Form>
            )}
        </Formik>
    </Container>)
}

InvitationEmailForm.propTypes = {
    onSuccess: Proptypes.func.isRequired,
}

export default InvitationEmailForm;