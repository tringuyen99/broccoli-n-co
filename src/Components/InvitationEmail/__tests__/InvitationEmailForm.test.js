import React from 'react';
import renderer from 'react-test-renderer';
import {render, screen, waitFor} from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import InvitationEmailForm from '../InvitationEmailForm';

import { apiFetch } from '../../../Utils/apiFetch';
import { invitationEmailAPI } from '../../../Utils/apis';

jest.mock('../../../utils/apiFetch');

const onSuccessStub = jest.fn();

const propsStub = {
    onSuccess: onSuccessStub,
};

describe('InvitationEmailForm', () =>{
    beforeAll(()=> {
        apiFetch.mockImplementation(
            (url, method, formData) => new Promise(resolve => resolve())
        );
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(<InvitationEmailForm {...propsStub} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('makes network request when the fields are correct', async () => {
        render(<InvitationEmailForm {...propsStub} />)

        apiFetch.mockImplementation(
            (url, method, formData) => new Promise(resolve => resolve())
        );
        
        userEvent.type(screen.getByPlaceholderText('Full Name'), 'John');
        userEvent.type(screen.getByPlaceholderText('Email'), 'john.dee@someemail.com');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john.dee@someemail.com');

        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: 'john.dee@someemail.com',
            name: 'John',
            }),
        );
        expect(onSuccessStub).toBeCalled();
    });

    it('displays error when the name is not given', async () => {
        render(<InvitationEmailForm {...propsStub} />)

        userEvent.type(screen.getByPlaceholderText('Email'), 'john.dee@someemail.com');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john.dee@someemail.com');
        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).not.toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: 'john.dee@someemail.com',
            name: '',
            }),
        );
        expect(await screen.findByText("*Name is required")).toBeInTheDocument();
        expect(onSuccessStub).not.toBeCalled();
    });

    it('displays error when the name is not long enough', async () => {
        render(<InvitationEmailForm {...propsStub} />)

        userEvent.type(screen.getByPlaceholderText('Full Name'), 'Jo');
        userEvent.type(screen.getByPlaceholderText('Email'), 'john.dee@someemail.com');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john.dee@someemail.com');
        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).not.toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: 'john.dee@someemail.com',
            name: 'Jo',
            }),
        );
        expect(await screen.findByText("*Names must have at least 3 characters")).toBeInTheDocument();
        expect(onSuccessStub).not.toBeCalled();
    });

    it('displays error when the email is not given', async () => {
        render(<InvitationEmailForm {...propsStub} />)

        userEvent.type(screen.getByPlaceholderText('Full Name'), 'John');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john.dee@someemail.com');
        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).not.toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: '',
            name: 'John',
            }),
        );
        expect(await screen.findByText("*Email is required")).toBeInTheDocument();
        expect(onSuccessStub).not.toBeCalled();
    });

    it('displays error when the email is not valid', async () => {
        render(<InvitationEmailForm {...propsStub} />)
        userEvent.type(screen.getByPlaceholderText('Full Name'), 'John');
        userEvent.type(screen.getByPlaceholderText('Email'), 'john.deel.com');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john.deel.com');
        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).not.toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: 'john.deel.com',
            name: 'John',
            }),
        );

        const errorsMessage = await screen.findAllByText("*Must be a valid email address");
        expect(errorsMessage[0]).toBeInTheDocument();
        expect(errorsMessage[1]).toBeInTheDocument();
        expect(onSuccessStub).not.toBeCalled();
    });

    it('displays error when the emails are not matched', async () => {
        render(<InvitationEmailForm {...propsStub} />)
        userEvent.type(screen.getByPlaceholderText('Full Name'), 'John');
        userEvent.type(screen.getByPlaceholderText('Email'), 'john.dee@someemail.com');
        userEvent.type(screen.getByPlaceholderText('Confirm Email'), 'john2.dee@someemail.com');
        userEvent.click(screen.getByRole('button'));

        await waitFor(() =>
            expect(apiFetch).not.toHaveBeenCalledWith(invitationEmailAPI, 'POST', {
            email: 'john.deel.com',
            name: 'John',
            }),
        );

        expect( await screen.findByText("*Emails must match")).toBeInTheDocument();
        expect(onSuccessStub).not.toBeCalled();
    });
})
