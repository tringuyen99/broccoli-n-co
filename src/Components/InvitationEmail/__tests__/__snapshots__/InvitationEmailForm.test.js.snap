// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitationEmailForm renders correctly 1`] = `
<div
  className="justify-content-center text-center container-fluid"
>
  <h1
    className="display-6 fw-bold"
  >
    Please fill in this form
  </h1>
  <p
    className="fw-light"
  >
    Broccoli&Co
  </p>
  <form
    className="mx-auto"
    onSubmit={[Function]}
  >
    <div
      className="mb-4"
    >
      <label
        className="form-label"
        htmlFor="formName"
      >
        Name :
      </label>
      <input
        className="border border-dark form-control"
        id="formName"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Full Name"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-4"
    >
      <label
        className="form-label"
        htmlFor="formEmail"
      >
        Email :
      </label>
      <input
        className="border border-dark form-control"
        id="formEmail"
        name="email"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
    </div>
    <div
      className="mb-5"
    >
      <label
        className="form-label"
        htmlFor="formConfirmEmail"
      >
        Confirm Email :
      </label>
      <input
        className="border border-dark form-control"
        id="formConfirmEmail"
        name="confirmEmail"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Confirm Email"
        type="text"
        value=""
      />
    </div>
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-primary btn-lg"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
